### PR TITLE
Flyte CoPilot: Raw container support for all K8s containers

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This PR implements FlyteCoPilot as an optional for all container executions. This is an _Alpha feature_

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?
 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Refer to [Design Doc](https://docs.google.com/document/d/1ZsCDOZ5ZJBPWzCNc45FhNtYQOxYHz0PAu9lrtDVnUpw/edit#)

## Tracking Issue
https://github.com/lyft/flyte/issues/297

## Follow-up issue
https://github.com/lyft/flyte/issues/316
https://github.com/lyft/flyte/issues/317

### CODE REVIEW

File: go/cmd/data/cmd/containerwatcher/spns_watcher.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	currentProcessPid int
+	pidsToFilter      sets.Int
+}
+
+func (s *SharedNamespaceProcessLister) AnyProcessRunning(ctx context.Context) (bool, error) {
+	procs, err := s.ListRunningProcesses(ctx)
+	if err != nil {
+		return false, err
+	}
+	return len(procs) > 0, nil
```
Comment (EngHabu) [inline go/cmd/data/cmd/containerwatcher/spns_watcher.go:47]: with "me" as a process, this will always be at least 1, right? shouldn't the check be `> 1` ?

File: go/cmd/data/cmd/containerwatcher/kubeapi_watcher.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+			return ErrTimeout
+		}
+	}
+}
+
+func (k kubeAPIWatcher) WaitToStart(ctx context.Context) error {
+	f := func(ctx context.Context, p *v13.Pod) (bool, error) {
+		for _, c := range p.Status.ContainerStatuses {
+			if c.Name == k.info.Name {
+				if c.Ready {
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/cmd/data/cmd/containerwatcher/kubeapi_watcher.go:55]: hmm I guess you are saying that in case we are pending in imagepull or we have failed then we should exit too, just with a failure exit code. Suggest If I should just remove this option for now?

File: go/cmd/data/cmd/containerwatcher/spns_watcher.go (hunk 1 above)
Comment (kumare3) [reply go/cmd/data/cmd/containerwatcher/spns_watcher.go:47]: no it removes the `/pause` and `self`. Check the documentation on Filter processe

File: go/cmd/data/cmd/containerwatcher/file_watcher.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
+package containerwatcher
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/cmd/data/cmd/containerwatcher/file_watcher.go:1]: ya it should be containerwatcher/file?

File: go/cmd/data/cmd/containerwatcher/kubeapi_watcher.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kumare3) [reply go/cmd/data/cmd/containerwatcher/kubeapi_watcher.go]: I am actually not using this mode at all right now.

File: go/tasks/pluginmachinery/flytek8s/copilot.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+			// Lets add the Inputs init container
+			args, err := DownloadCommandArgs(inputPaths.GetInputPath(), outputPaths.GetOutputPrefixPath(), inPath, format, iFace.Inputs)
+			if err != nil {
+				return err
+			}
+			downloader, err := FlyteCoPilotContainer(flyteInitContainerName, cfg, args, inputsVolumeMount, cfgVMount)
+			if err != nil {
+				return err
+			}
+			coPilotPod.InitContainers = append(coPilotPod.InitContainers, downloader)
```
Comment (wild-endeavor) [inline go/tasks/pluginmachinery/flytek8s/copilot.go:234]: Appending will make it the last init container - should we make it the first instead?

File: copilot/cmd/root.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	return r.Store.WriteProtobuf(ctx, errorPath, storage.Options{}, &core.ErrorDocument{
+		Error: &core.ContainerError{
+			Code:    code,
+			Message: recvErr.Error(),
+			Kind:    core.ContainerError_RECOVERABLE,
+		},
+	})
+}
+
+// NewCommand returns a new instance of an argo command
```
Comment (wild-endeavor) [inline copilot/cmd/root.go:69]: argo command?

File: copilot/cmd/sidecar.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		logger.Errorf(ctx, "Uploading failed, err %s", err)
+		if err := u.UploadError(ctx, "OutputUploadFailed", err, storage.DataReference(u.remoteOutputsPrefix)); err != nil {
+			logger.Errorf(ctx, "Failed to write error document, err :%s", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func NewUploadCommand(opts *RootOptions) *cobra.Command {
```
Comment (wild-endeavor) [inline copilot/cmd/sidecar.go:155]: In general, I feel like the uploader, downloader, and sidecar watcher should be different cmds in flytectl.  That could get messy too but maybe worth thinking about?

File: Makefile [hunk 7]
```diff
@@ -10,3 +11,17 @@ generate: download_tooling
...
+linux_compile:
+	cd copilot; GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build -o ../artifacts/flyte-copilot .; cd -
+
+.PHONY: compile
+compile:
+	mkdir -p ./artifacts
+	cd copilot; go build -o ../artifacts/flyte-copilot .; cd -
+
+cross_compile:
+	@glide install
```
[the remark this answers is not in the document]
Comment (kumare3) [reply Makefile:25]: what good catch, how did that get in there?

[the remark this answers is not in the document]
Comment (kumare3) [reply Makefile:25]: pre go.mod go i think

File: copilot/cmd/download.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+func GetUploadModeVals() []string {
+	var vals []string
+	for k := range core.IOStrategy_UploadMode_value {
+		vals = append(vals, k)
+	}
+	return vals
+}
+
+func (d *DownloadOptions) Download(ctx context.Context) error {
+	if d.remoteOutputsPrefix == "" {
```
[the remark this answers is not in the document]
Comment (kumare3) [reply copilot/cmd/download.go:52]: yes metrics i want to add, but as a second pass

File: copilot/data/download.go [hunk 9]
```diff
@@ -0,0 +1,365 @@
+package data
```
[the remark this answers is not in the document]
Comment (kumare3) [reply copilot/data/download.go:1]: You have absolutely hit a point that I was thinking as well. CoPilot should be part of flytectl?

File: go/tasks/pluginmachinery/utils/literals.go [hunk 10]
```diff
@@ -0,0 +1,439 @@
+package utils
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/tasks/pluginmachinery/utils/literals.go:1]: yes after this gets merged i was going to move this to there and move eventing etc into propeller

File: go/tasks/pluginmachinery/flytek8s/config/config.go [hunk 11]
```diff
@@ -17,6 +25,21 @@ var (
 		DefaultAnnotations: map[string]string{
 			"cluster-autoscaler.kubernetes.io/safe-to-evict": "false",
 		},
+		CoPilot: FlyteCoPilotConfig{
+			NamePrefix:           "flyte-copilot-",
+			Image:                "docker.pkg.github.com/lyft/flyteplugins/operator:v0.4.0",
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/tasks/pluginmachinery/flytek8s/config/config.go:30]: no, its a config, so you can just change the config for the deployment?

File: go/tasks/pluginmachinery/flytek8s/copilot.go (hunk 4 above)
Comment (kumare3) [reply go/tasks/pluginmachinery/flytek8s/copilot.go:234]: cannot make the first like in case of IAM wait, but IAM wait is injected. Also, we do not really add the data volume to all init containers, thus not sure, if making it earlier has any uses?

File: copilot/cmd/root.go (hunk 5 above)
Comment (kumare3) [reply copilot/cmd/root.go:69]: haha good catch, not sure what my train of thought was?

File: copilot/cmd/sidecar.go (hunk 6 above)
Comment (kumare3) [reply copilot/cmd/sidecar.go:155]: no this is not messy at all. I think I agree, I think they should be in flytectl, just that we need to move some code to flyteidl before we do that. What do you think we move ahead and then clean up after the fact. I can drop all the github workflows?

File: copilot/data/future.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
+// This module implements a simple Async Futures for golang
```
[the remark this answers is not in the document]
Comment (kumare3) [reply copilot/data/future.go:1]: :). I will move it to flytestdlib as a second pass

File: copilot/data/download.go [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/lyft/flytestdlib/logger"
+	"github.com/lyft/flytestdlib/storage"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+type Downloader struct {
+	format core.DataLoadingConfig_LiteralMapFormat
+	// TODO support multiple buckets
```
[the remark this answers is not in the document]
Comment (kumare3) [reply copilot/data/download.go:27]: ohh never mind i added support for multiple buckets already, removing the comment

File: copilot/cmd/sidecar.go (hunk 6 above)
Comment (wild-endeavor) [reply copilot/cmd/sidecar.go:155]: sure sure, sounds good.

**Comment by wild-endeavor** [comment]
just let me know when you want the final +1

File: copilot/cmd/download.go [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+		return fmt.Errorf("to-remoute-prefix is required")
```
[the remark this answers is not in the document]
Comment (kumare3) [reply copilot/cmd/download.go:53]: Done
